### DISCUSSION TITLE
Embedding dynamic (i.e. non-Quarto YAML) variables/outputs into 'vanilla' Markdown

### DISCUSSION THREAD

**Discussion by jreades** [quarto-dev/quarto-cli#8287] [Q&A] [answered]
### Description

I can't find any examples/discussions relevant to this even though I struggle a bit to imagine that I'm the first person to want to do this... Anyway, I'd like to make it simpler to embed output from code executed as _part_ of the rendering process into the final output. So a dynamic version of the `{{< var var_name >}}` such that a Quarto document creator can write something like: 

``````python
```{python}
tst = 'Hello World'
```
``````

And then have straightforward Markdown along the lines of:

> I'd like to be able to embed `{{ test }}` (or `{{< test >}}`) into some vanilla **Markdown**.

And this would be output as:

> I'd like to be able to embed `Hello World` (or `Hello World`) into some vanilla **Markdown**.

I'd originally thought this would be a post-render script in which variables to be embedded were 'collected' into a key/value store and then embedded post-render. That's the least 'invasive' way to do it and could operate independently from the programming language(s) used in the QMD. So it's the most predictable, but it _also_ relies on a clunky store-during-render/retrieve-after-then-embed process. I suppose something like `{{< python tst >}}` (i.e. get the `tst` variable stored by the Python interpreter) and you'd look in the tmp file for `python:<var name>`?

So the more I dug into it the more I began to *suspect* that it might be neater to handle it as a *pre*-render such that the QMD file is scanned and any Markdown header/paragraph containing an unescaped `{{ ... }}` (though you'd need to look out for Quarto's own embeds!) would be converted to the output of a function along these lines:

```{python}
from IPython.display import display, Markdown

tst = 'Hello World'
number = 123.457

display(
    Markdown(
    """
### Some Header

We find that it's possible to output '{tst}' text as well as {number:.2f}, but only with a fair bit of faff.
""".format(tst=tst, number=number)
   )
)
```

*Aside:* this example taken from [here](https://aeturrell.github.io/coding-for-economists/wrkflow-quarto.html)

Of course, since QMD files can contain multiple languages, perhaps this needs to be in the format `{{< python number:.2f >}}` or `{{ number }}{=python}` or something similar. So the pre-render keeps getting messier and messier, but the idea of writing my documents _as above_ using the `display( Markdown( ... ) )` syntax is... anathema to me when there _should_ be an elegant way to do it and clearly Quarto itself has a mechanism for doing this because of the `var` syntax.

Hope this all makes sense!

**Comment by mcanouil** [comment, marked as answer]
Don't use `\`, use more backticks to enclose code with code blocks/cells.

> non-Quarto YAML

What is that?

YAML metadata block are delimited with `---`.
If your code generate something like `\n---\nmy-variable: amazing\n---\n`, then you can could the meta shortcode `{{< meta my-variable >}}`.

If you intend to use `{{< meta my-variable >}}` inside code cells, you can't.
you need to use parameters as described in the documentation:
- https://quarto.org/docs/authoring/variables.html#meta
- https://quarto.org/docs/computations/parameters.html

For more, plase provide a complete "working" example:


`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
#| output: asis
cat("\n---\nmy-variable: amazing\n---\n")
```

![A placeholder image](https://placehold.co/600x400.png)

The end of the demo using {{< meta my-variable >}} stuff.
````
`````

**Reply by jreades** [reply]
Aaaaah, I had been *way* over-thinking this: partly because I forgot how QMD documents are rendered but also because I didn't know that inline YAML was an option! My need wasn't using parameters in code, just using the results of a calculation (e.g. the mean of an attribute in a data set) as part of a paragraph of text (e.g. "From this we can see that the mean of attribute X ({{< meta value >}}) is above what we'd expect from the findings of @author:year, implying that..."). This approach still gets a little clunky if you *really* want to dynamically fill out the report (e.g. inserting 'greater than' or 'less than' dynamically) but this makes life much, much easier than I'd expected/realised based on what I'd found in my Google searches.

*Anyway*, after a bit of playing I've worked out the Python equivalent as:

``````cmd
---
title: "Variable Test"
format:
  html:
    toc: false
---

```{python}
tst = 'Hello world(s)'
number = 1000.2347
```

```{python}
#| output: asis
#| echo: true

# Option 1
#print(f"\n---\nbar: '{tst}'\n---\n")
#print(f"\n---\nbaz: '{number:.2f}'\n---\n")

# Option 2
print(f"---")
print(f"bar: '{tst}'")
print(f"baz: '{number:.2f}'")
print(f"---")
```

How about **{{< meta bar >}}** and *{{< meta baz >}}*?
``````

Thank you for the quickly reply! This use-case *might* be a useful addition to the tail-end of the documentation on variables since it's not a natural/logical extension of what is there on the use of external YAML files and nor is it immediately obvious that these will become `meta` parameters rather than `var` parameters.

**Reply by jreades** [reply]
Just to tidy up the last unanswered question: by 'Non-Quarto YAML' I meant a YAML *file* that wasn't one of the three specified in the documentation (`var`, `meta`, `env`). That was *all* before I realised you could write in additional YAML data outside of the header. In retrospect it makes sense (i.e. you see the Python bits execute before anything else happens so it's now obvious to me that you *could* add meta YAML to the output programmatically) but it's not an easy model to get your head around when it's not part of your day-to-day writing: I'd always thought this content was kind of 'fixed' once rendering started.

Also, my frame of reference here was stuff like `mod_perl` and PHP where you have a syntax for inserting a variable from the current 'run' directly into the text. They are actually doing the *reverse* of what Quarto does in many ways, so that's where I went down the rabbit hole.

**Reply by mcanouil** [reply]
We currently don't want to advertise the multi-YAML block too much because it hasn't been fully tested against everything in Quarto.
It's a Pandoc feature which happens to work because computations are nearly the first thing done.
Note that this will work as expected for all "meta you create" that does not conflict with Quarto/Pandoc, so I would suggest for you to create a scheme such as it is safe.

**Reply by cscheid** [reply]
@mcanouil We're actually going to add documentation about this in the near future.

**Reply by mcanouil** [reply]
I know, but not right now, so I prefer to add a note of caution here. ;)

**Comment by cderv** [comment]
> Of course, since QMD files can contain multiple languages, perhaps this needs to be in the format {{< python number:.2f >}} or {{ number }}{=python} or something similar.

It seems you are satisfied with the other answer, but I believe you are looking for Inline execution for Jupyter. 
https://quarto.org/docs/prerelease/1.4/inline.html


This is a new feature of Quarto 1.4. Your document would look like 

````markdown
---
title: "Variable Test"
format:
  html:
    toc: false
---

```{python}
tst = 'Hello world(s)'
number = 1000.2347
```

How about `{python} tst` and `{python} f"{number:.2f}"` ?

````

![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/1283c327-da40-4689-bb1a-c3c67bd4ddda)


This is the way to use some computed value in Markdown text directly - no need to rely on YAML creation and meta usage to pass values. 

Would that work for you ?

**Reply by cderv** [reply]
Seeing bold and italic  now in your other example, so adding one more 
````markdown
---
title: "Variable Test"
format:
  html:
    toc: false
---

```{python}
from IPython.display import Markdown
tst = 'Hello world(s)'
number = 1000.2347
```

How about `{python} Markdown(f"**{tst}**")` and `{python} Markdown(f"*{number:.2f}*")` ?


````

![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/1d94678a-acc0-4ef6-947e-db09a8ecd503)

**Reply by jreades** [reply]
Thank you, that's really helpful. I can see pros and cons to each approach depending on the use case, so it's really useful to have this flexibility and it's radically simpler than what I'd imagined. So 1.4 stable is due... (pessimistically) by summer? I can live with that -- I am happy to work with pre-release for my own work but would prefer to teach the official release.

**Reply by mcanouil** [reply]
1.4 is in "release candidate" state meaning only minor changes should occur now and that it is the main download version on Quarto website.
Release is expected to be soon.
